ts_src/payments/p2tr.ts

```diff
@@ -0,0 +1,349 @@
+import { Buffer as NBuffer } from 'buffer';
+import { bitcoin as BITCOIN_NETWORK } from '../networks';
+import * as bscript from '../script';
+import { typeforce as typef, isTaptree, TAPLEAF_VERSION_MASK } from '../types';
+import { getEccLib } from '../ecc_lib';
+import {
+  toHashTree,
+  rootHashFromPath,
+  findScriptPath,
+  tapleafHash,
+  tapTweakHash,
+  LEAF_VERSION_TAPSCRIPT,
+} from './taprootutils';
+import { Payment, PaymentOpts } from './index';
+import * as lazy from './lazy';
+import { bech32m } from 'bech32';
+
+const OPS = bscript.OPS;
+const TAPROOT_WITNESS_VERSION = 0x01;
+const ANNEX_PREFIX = 0x50;
+
+export function p2tr(a: Payment, opts?: PaymentOpts): Payment {
+  if (
+    !a.address &&
+    !a.output &&
+    !a.pubkey &&
+    !a.internalPubkey &&
+    !(a.witness && a.witness.length > 1)
+  )
+    throw new TypeError('Not enough data');
+
+  opts = Object.assign({ validate: true }, opts || {});
+
+  typef(
+    {
+      address: typef.maybe(typef.String),
+      input: typef.maybe(typef.BufferN(0)),
+      network: typef.maybe(typef.Object),
+      output: typef.maybe(typef.BufferN(34)),
+      internalPubkey: typef.maybe(typef.BufferN(32)),
+      hash: typef.maybe(typef.BufferN(32)), // merkle root hash, the tweak
+      pubkey: typef.maybe(typef.BufferN(32)), // tweaked with `hash` from `internalPubkey`
+      signature: typef.maybe(typef.BufferN(64)),
```

Comment:
Can be 64 or 65 bytes.

Reply:
It looks like [option 3 was chosen](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#design), has something changed?


> Encoding R and public key point P There exist several possibilities for encoding elliptic curve points:
> 
>     1. Encoding the full X and Y coordinates of P and R...
>     2. Encoding the full X coordinate and one bit of the Y...
>     3. Encoding only the X coordinate, resulting in 32-byte public keys and 64-byte signatures.
> 
> Using the first option would be slightly more efficient for verification (around 10%), but we prioritize compactness, and therefore choose option 3. 

Also in the [bitcoin core secp256k1](https://github.com/bitcoin/bitcoin/blob/master/src/secp256k1/include/secp256k1_schnorrsig.h#L110):

```c
/** Create a Schnorr signature.
* ...
*   Out:   sig64: pointer to a 64-byte array to store the serialized signature.
SECP256K1_API int secp256k1_schnorrsig_sign(
```

Reply:
It can include an optional `SIGHASH_TYPE`. A 64-byte signature encodes `SIGHASH_DEFAULT` => `0x00`. A 65-byte signature encodes an explicit `SIGHASH_TYPE`.

Reply:
this should be included as a precondition for the `isCanonicalSchnorrSignature` check